### PR TITLE
[PBI-1824] Add SSAI Pulse Plugin

### PR DESCRIPTION
**Pull request by dchauOoyala**

### CODE REVIEW

File: js/ssai_pulse.js [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+      amc.updateMainStreamUrl(url);
+      baseRequestUrl = _preformatUrl(baseRequestUrl);
+    };
+
+    /**
+     * This is an example callback that interprets video stream tags.  The event is subscribed to in
+     * the initialize function.
+     * @public
+     * @method SsaiPulse#onVideoTagFound
+     * @param {string} event The event that triggered this callback
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: jsdoc doesn't match parameter

File: js/ssai_pulse.js [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+    /**
+     * Helper function to append "offset" and "aid" query parameters to the request URL.
+     * @private
+     * @method SsaiPulse#_appendAdsProxyQueryParameters
+     * @param {string} url The request URL
+     * @param {string} adId The ID of the ad
+     * @returns {string} The request URL with the appended query parameters.
+     */
+    var _appendAdsProxyQueryParameters = function(url, adId) {
+      // vastUrl + '&offset=5&aid=' + adid
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: there are already initial query parameters in the URL at this point

File: js/ssai_pulse.js [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+          if (this.currentAd) {
+            _adEndedCallback();
+          }
+
+          // Will call _sendRequest() once live team fixes ads proxy issue. Will directly call onResponse() for now.
+          if (!this.testMode) {
+            //_sendRequest(requestUrl);
+            this.onResponse(null, this.currentId3Object.duration);
+
+            // Set timer for duration of the ad.
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: put this above request

File: js/ssai_pulse.js [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+
+    /**
+     * Called if the ajax call succeeds
+     * @public
+     * @method SsaiPulse#onResponse
+     * @param {XMLDocument} xml The xml returned from loading the ad
+     * @param {number} adDuration The duration of the current ad
+     */
+    this.onResponse = function(xml, adDuration) {
+      console.log("SSAI Pulse: Response");
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: OO.log

File: js/ssai_pulse.js [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+
+    /**
+     * Callback used when the duration of an ad has passed.
+     * @private
+     * @method SsaiPulse#_adEndedCallback
+     */
+    var _adEndedCallback = _.bind(function() {
+      amc.notifyLinearAdEnded(this.currentAd.id);
+      amc.notifyPodEnded(this.currentAd.id);
+      _clearAdDurationTimeout();
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: clear ad timeout first

File: test/unit-tests/ssai_pulse_test.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    mockId3Tag.TXXX = "";
+    expectedResult = null;
+    SsaiPulse.onVideoTagFound("eventName", "videoId", "tagType", mockId3Tag);
+    expect(SsaiPulse.currentId3Object).to.be(null);
+
+    mockId3Tag.TXXX = null;
+    expectedResult = null;
+    SsaiPulse.onVideoTagFound("eventName", "videoId", "tagType", mockId3Tag);
+    expect(SsaiPulse.currentId3Object).to.be(null);
+  });
```
Comment (dchauOoyala) [inline test/unit-tests/ssai_pulse_test.js]: Add unit test to test ad ending with new ID3 tag

File: js/ssai_pulse.js [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+    // Tracking Event states
+    var adMode = false;
+    var isFullscreen = false;
+    var isMuted = false;
+    var lastVolume = -1;
+
+    // Params required for ads proxy ads request
+    var SMART_PLAYER = "&oosm=1";
+    var OFFSET_PARAM = "&offset=";
+    var OFFSET_VALUE = "5";
```
Comment (dchauOoyala) [inline js/ssai_pulse.js]: what unit

File: js/ssai_pulse.js [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+     *                                        started
+     * @param {function} adPodEndedCallback Call this function when the ad or group of podded ads have
+     *                                      completed
+     * @param {function} adStartedCallback Call this function each time an ad in the set starts
+     * @param {function} adEndedCallback Call this function each time an ad in the set completes
+     */
+    this.playAd = function(ad, adPodStartedCallback, adPodEndedCallback, adStartedCallback, adEndedCallback) {
+      adMode = true;
+      this.currentAd = ad;
+      amc.notifyLinearAdStarted(ad.id, {
```
Comment (pilievOoyala) [inline js/ssai_pulse.js]: we need to check that ad and ad.ad exist

File: js/ssai_pulse.js [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+
+    /**
+     * <i>Optional.</i><br/>
+     * Called when player clicks on the tap frame, if tap frame is disabled, then this function will not be
+     * called
+     * @method SsaiPulse#playerClicked
+     * @public
+    */
+    this.playerClicked = function(amcAd, showPage) {
+      window.open(amcAd.ad.clickthrough);
```
Comment (pilievOoyala) [inline js/ssai_pulse.js]: i know this probably came from the template but need to check that amcAd and amcAd.ad exist

File: js/ssai_pulse.js [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+
+    /**
+     * Appends the smart player identifier to the request URL.
+     * @private
+     * @method SsaiPulse#_makeSmartUrl
+     * @param {string} url The stream url
+     * @returns {string} The modified stream url with the appended unique identifier.
+     */
+    var _makeSmartUrl = function(url) {
+      return url + SMART_PLAYER;
```
Comment (pilievOoyala) [inline js/ssai_pulse.js]: don't need to do this now, but maybe use the jQuery url builder instead.

File: js/ssai_pulse.js [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+     * the Ad Manager Controller.
+     * @method SsaiPulse#destroy
+     * @public
+     */
+    this.destroy = function() {
+      // Stop any running ads
+      this.ready = false;
+      this.currentAd = null;
+      this.currentId3Object = null;
+      adIdDictionary = {};
```
Comment (pilievOoyala) [inline js/ssai_pulse.js]: need to add in amc.removeEventListener to all the messages you add.

**Comment by pilievOoyala** [comment]
Yay! LGTM.

**Pull request merged**